File: kompile-app/kompile-process-engine/src/main/java/ai/kompile/process/service/ProcessEngineService.java
/*
 *   Copyright 2025 Kompile Inc.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ai.kompile.process.service;

import ai.kompile.process.controls.ControlAttestation;
import ai.kompile.process.controls.ControlDefinition;
import ai.kompile.process.execution.WorkflowRun;
import ai.kompile.process.hitl.ApprovalRequest;
import ai.kompile.process.hitl.ApprovalResponse;
import ai.kompile.process.ingest.SubmissionManifest;
import ai.kompile.process.ontology.OntologySchema;
import ai.kompile.process.ontology.ValidationRule;
import ai.kompile.process.workflow.ProcessDefinition;

import java.util.List;
import java.util.Map;

/**
 * Main service interface for the process engine.
 * Orchestrates workflow execution, HITL gates, and control evaluation.
 * Implementations of this interface drive the full lifecycle from ontology design
 * through process execution and SOX-compliant audit trail generation.
 */
public interface ProcessEngineService {

    // -------------------------------------------------------------------------
    // Ontology management
    // -------------------------------------------------------------------------

    /**
     * Persists a new ontology schema and returns it with a generated ID and version 1.
     *
     * @param schema the schema to create
     * @return the created schema with server-assigned ID and metadata
     */
    OntologySchema createOntology(OntologySchema schema);

    /**
     * Retrieves a specific version of an ontology schema.
     *
     * @param id      the ontology schema ID
     * @param version the version number to retrieve
     * @return the requested ontology schema snapshot
     */
    OntologySchema getOntology(String id, int version);

    /**
     * Updates an existing ontology schema, incrementing the version number.
     * Previous versions remain immutable and available for retrieval.
     *
     * @param id     the ID of the schema to update
     * @param schema the new schema content
     * @return the updated schema with an incremented version
     */
    OntologySchema updateOntology(String id, OntologySchema schema);

    /**
     * Validates a data map against the rules defined for a specific entity type
     * in the given ontology version.
     *
     * @param ontologyId  the ontology schema ID
     * @param version     the ontology version to use for validation
     * @param entityType  the entity type name whose rules to evaluate
     * @param data        the key-value data to validate
     * @return a list of violated {@link ValidationRule}s (empty if all rules pass)
     */
    List<ValidationRule> validateData(String ontologyId, int version, String entityType, Map<String, Object> data);

    // -------------------------------------------------------------------------
    // Process definition
    // -------------------------------------------------------------------------

    /**
     * Creates a new process definition in DRAFT status.
     *
     * @param definition the process definition to persist
     * @return the created definition with server-assigned ID and version 1
     */
    ProcessDefinition createProcess(ProcessDefinition definition);

    /**
     * Retrieves a specific version of a process definition.
     *
     * @param id      the process definition ID
     * @param version the version number to retrieve
     * @return the requested process definition snapshot
     */
    ProcessDefinition getProcess(String id, int version);

    /**
     * Transitions a process definition from IN_REVIEW to APPROVED.
     * Approved definitions may be used to start workflow runs.
     *
     * @param id         the process definition ID to approve
     * @param approvedBy the identifier of the person granting approval
     * @return the updated definition with status APPROVED
     */
    ProcessDefinition approveProcess(String id, String approvedBy);

    /**
     * Returns all ontology schemas (latest version of each).
     */
    List<OntologySchema> listOntologies();

    /**
     * Returns all process definitions (latest version of each).
     */
    List<ProcessDefinition> listProcessDefinitions();

    // -------------------------------------------------------------------------
    // Workflow execution
    // -------------------------------------------------------------------------

    /**
     * Starts a new workflow run for the specified process definition.
     * The process must have APPROVED status before a run can be started.
     *
     * @param processDefinitionId the ID of the approved process to execute
     * @param initialData         seed data injected into the run's data state before execution begins
     * @return the newly created {@link WorkflowRun} in RUNNING status
     */
    WorkflowRun startRun(String processDefinitionId, Map<String, Object> initialData);

    /**
     * Retrieves the current state of a workflow run.
     *
     * @param runId the workflow run ID
     * @return the current {@link WorkflowRun} snapshot
     */
    WorkflowRun getRun(String runId);

    /**
     * Resumes a run that is paused at a HITL approval gate after a human response has been submitted.
     *
     * @param runId    the ID of the paused run to resume
     * @param response the human's approval response
     * @return the updated {@link WorkflowRun} after processing the response
     */
    WorkflowRun resumeAfterApproval(String runId, ApprovalResponse response);

    /**
     * Lists all workflow runs that are currently in an active state
     * (RUNNING, PAUSED_FOR_APPROVAL, or PAUSED_FOR_HUMAN).
     *
     * @return a list of active {@link WorkflowRun} instances
     */
    List<WorkflowRun> listActiveRuns();

    /**
     * Returns all workflow runs regardless of status.
     */
    List<WorkflowRun> listAllRuns();

    /**
     * Cancels a running or paused workflow run.
     *
     * @param runId the ID of the run to cancel
     * @return the updated run with CANCELLED status
     */
    WorkflowRun cancelRun(String runId);

    /**
     * Completes a HUMAN-type step that is waiting for manual execution.
     *
     * @param runId       the workflow run ID
     * @param stepId      the step ID to complete
     * @param completedBy who completed the step
     * @param outputs     the outputs produced by the human
     * @return the updated run after advancing past the completed step
     */
    WorkflowRun completeHumanStep(String runId, String stepId, String completedBy, Map<String, Object> outputs);

    // -------------------------------------------------------------------------
    // Human-in-the-loop (HITL)
    // -------------------------------------------------------------------------

    /**
     * Returns all pending approval requests assigned to the specified person or role.
     *
     * @param assignedTo the person ID or role name to filter by
     * @return a list of {@link ApprovalRequest}s awaiting response
     */
    List<ApprovalRequest> getPendingApprovals(String assignedTo);

    /**
     * Retrieves a single approval request by ID.
     *
     * @param requestId the approval request ID
     * @return the {@link ApprovalRequest}
     */
    ApprovalRequest getApprovalRequest(String requestId);

    /**
     * Submits a human response to an approval request and resumes the associated workflow run.
     *
     * @param response the human's approval response
     * @return the updated {@link WorkflowRun} after the response is processed
     */
    WorkflowRun submitApproval(ApprovalResponse response);

    // -------------------------------------------------------------------------
    // Controls
    // -------------------------------------------------------------------------

    /**
     * Evaluates a control assertion against the current data state of a run.
     * Always produces an immutable {@link ControlAttestation} record regardless of outcome.
     *
     * @param controlId the ID of the control definition to evaluate
     * @param runId     the workflow run ID providing data context
     * @param data      the data values to evaluate the control expression against
     * @return the immutable attestation record
     */
    ControlAttestation evaluateControl(String controlId, String runId, Map<String, Object> data);

    /**
     * Returns all control attestation records produced during a workflow run.
     *
     * @param runId the workflow run ID
     * @return an ordered list of {@link ControlAttestation} records
     */
    List<ControlAttestation> getControlResults(String runId);

    /**
     * Creates a new control definition at runtime.
     */
    ControlDefinition createControl(ControlDefinition control);

    /**
     * Returns all registered control definitions.
     */
    List<ControlDefinition> listControls();

    /**
     * Retrieves a single control definition by ID.
     */
    ControlDefinition getControl(String controlId);

    /**
     * Evaluates a SpEL expression against a context map without requiring a saved ControlDefinition.
     *
     * @param expression the SpEL expression string
     * @param context    variable bindings for the evaluation context
     * @return the evaluation result including the value, type, and any error
     */
    SpelEvaluationResult evaluateSpelExpression(String expression, Map<String, Object> context);

    // -------------------------------------------------------------------------
    // Ingestion
    // -------------------------------------------------------------------------

    /**
     * Creates a submission manifest for a new batch of incoming data files.
     *
     * @param manifest the manifest to persist
     * @return the created manifest with a server-assigned ID
     */
    SubmissionManifest createManifest(SubmissionManifest manifest);

    /**
     * Records a human assertion that a specific file is the authoritative version
     * for a given submission cycle, resolving the "v3_FINAL_v2" ambiguity problem.
     *
     * @param manifestId  the manifest to update
     * @param fileId      the file ID being declared authoritative
     * @param assertedBy  the person making the assertion (name or user ID)
     * @return the updated manifest with the authoritative file recorded
     */
    SubmissionManifest assertAuthoritativeVersion(String manifestId, String fileId, String assertedBy);

    /**
     * Merges key-value data into the accumulated run data map of a workflow run.
     * Use this to store agent outputs, generated code artifacts, or any intermediate
     * results that other steps or tools should be able to access.
     *
     * @param runId the workflow run ID
     * @param data  key-value pairs to merge into the run's data map
     * @return the updated {@link WorkflowRun}
     */
    WorkflowRun updateRunData(String runId, Map<String, Object> data);

    // ════════════════ Excel Formula Conversion ════════════════

    /**
     * Converts Excel spreadsheet formulas to executable code via LLM.
     * Returns the generated code as an artifact for review and editing
     * without executing it.
     *
     * @param spreadsheetGraphJson  the SpreadsheetGraph JSON
     * @param targetLanguage        "javascript" or "python"
     * @return map containing "code", "language", "inputCells", "outputCells", metadata
     */
    Map<String, Object> convertExcelFormulas(String spreadsheetGraphJson, String targetLanguage);

    /**
     * Executes Excel spreadsheet computation. If generatedCode is provided,
     * it runs that code directly (user-edited artifact). Otherwise, converts
     * via LLM first. Result always includes _generatedCode for persistence.
     *
     * @param spreadsheetGraphJson  the SpreadsheetGraph JSON
     * @param cellOverrides         input cell value overrides
     * @param targetLanguage        "javascript" or "python"
     * @param generatedCode         user-supplied code (null to trigger LLM conversion)
     * @return map of computed outputs including _generatedCode artifact
     */
    Map<String, Object> executeExcelFormulas(String spreadsheetGraphJson,
                                              Map<String, Object> cellOverrides,
                                              String targetLanguage,
                                              String generatedCode);
}

File: kompile-app/kompile-app-main/src/main/frontend/src/app/services/process-engine.service.ts
/*
 * Copyright 2025 Kompile Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { BaseService } from './base.service';

// ─── Interfaces ───────────────────────────────────────────────────────────────

export interface OntologySchema {
  id?: string;
  name: string;
  version?: number;
  templateId?: string;
  createdAt?: string;
  updatedAt?: string;
  updatedBy?: string;
  entityTypes?: EntityTypeDefinition[];
  relationshipTypes?: RelationshipTypeDefinition[];
  globalRules?: ValidationRule[];
  metadata?: Record<string, any>;
}

export interface EntityTypeDefinition {
  name: string;
  description?: string;
  classification?: string;
  fields?: FieldDefinition[];
  rules?: ValidationRule[];
}

export interface FieldDefinition {
  name: string;
  type: string;
  required?: boolean;
  description?: string;
}

export interface RelationshipTypeDefinition {
  name: string;
  sourceEntityType: string;
  targetEntityType: string;
  cardinality?: string;
  description?: string;
}

export interface ValidationRule {
  name: string;
  type?: string;
  severity?: string;
  expression?: string;
  description?: string;
}

export interface ProcessDefinition {
  id?: string;
  name: string;
  version?: number;
  ontologySchemaId?: string;
  ontologyVersion?: number;
  status?: string;
  approvedBy?: string;
  approvedAt?: string;
  phases?: ProcessPhase[];
  controls?: ControlDefinitionRef[];
  agentSpecs?: string[];
  metadata?: Record<string, any>;
  /** ID of the parent process if this is a sub-process. */
  parentProcessId?: string;
  /** IDs of child (sub) processes nested within this process. */
  childProcessIds?: string[];
}

export interface ProcessPhase {
  id: string;
  name: string;
  order: number;
  steps?: ProcessStep[];
}

export interface ProcessStep {
  id: string;
  name: string;
  description?: string;
  stepType: 'AUTO' | 'APPROVE' | 'HUMAN' | 'CONTROL_GATE' | 'TOOL_CALL' | 'HTTP_CALL' | 'SCRIPT' | 'EXCEL_COMPUTE' | 'PIPELINE' | 'CAMEL_ROUTE' | 'DROOLS_RULE' | 'DROOLS_INFERENCE' | 'DROOLS_DECISION_TABLE' | 'WORKFLOW';
  inputKeys?: string[];
  outputKeys?: string[];
  agentSpecId?: string;
  approvalPolicy?: ApprovalPolicy;
  executionExpressions?: Record<string, string>;
  toolName?: string;
  toolArguments?: Record<string, string>;
  httpMethod?: string;
  httpUrl?: string;
  httpHeaders?: Record<string, string>;
  httpBody?: string;
  httpResponseKey?: string;
  scriptLanguage?: string;
  scriptBody?: string;
  scriptOutputKey?: string;
  excelGraphJson?: string;
  excelCellOverrides?: Record<string, any>;
  excelTargetLanguage?: string;
  excelOutputKey?: string;
  excelGeneratedCode?: string;
  pipelineDefinitionJson?: string;
  pipelineInputMapping?: Record<string, string>;
  pipelineOutputKey?: string;
  camelRouteId?: string;
  camelInlineScript?: string;
  camelOutputKey?: string;
  droolsRuleDrl?: string;
  droolsAgendaGroup?: string;
  droolsMaxFirings?: number;
  droolsDecisionTable?: string;
  droolsInputType?: 'CSV' | 'XLS' | 'XLSX';
  droolsWorksheetName?: string;
  droolsOutputKey?: string;
  droolsDecisionKey?: string;
  workflowEngineType?: 'xircuits' | 'n8n';
  workflowName?: string;
  workflowInlineContent?: string;
  workflowTimeoutSeconds?: number;
  workflowInputMapping?: Record<string, string>;
  workflowOutputKey?: string;
  controlIds?: string[];
  dependsOn?: string[];
  conditionExpression?: string;
  conditionLabel?: string;
  requiredRoles?: string[];
  requiredPermissions?: string[];
  graphNodeIds?: string[];
  confidence?: number;
  metadata?: Record<string, any>;
}

export interface ApprovalPolicy {
  approverPool?: string[];
  mode?: string;
  allowPartialApproval?: boolean;
  dollarThreshold?: number;
  maxAutoCorrections?: number;
}

export interface ControlDefinitionRef {
  controlId: string;
  triggerAfterStep: string;
}

export interface WorkflowRun {
  id?: string;
  processDefinitionId: string;
  processVersion?: number;
  status?: string;
  startedAt?: string;
  completedAt?: string;
  stepExecutions?: StepExecution[];
  pendingApprovals?: ApprovalRequest[];
  controlResults?: ControlAttestation[];
  runData?: Record<string, any>;
  graphNodeIds?: string[];
  metrics?: Record<string, any>;
}

export interface StepExecution {
  stepId: string;
  stepName?: string;
  status: string;
  startedAt?: string;
  completedAt?: string;
  executedBy?: string;
  inputs?: Record<string, any>;
  outputs?: Record<string, any>;
  inputHash?: string;
  outputHash?: string;
  evidenceReliedOn?: string[];
  graphNodeIds?: string[];
  error?: string;
}

export interface ControlDefinition {
  id?: string;
  name: string;
  description?: string;
  gateType?: 'HARD' | 'SOFT';
  expression?: string;
  triggerAfterStep?: string;
  inputKeys?: string[];
  regulatoryReference?: string;
  severity?: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
}

export interface ApprovalRequest {
  id?: string;
  workflowRunId?: string;
  stepId?: string;
  stepName?: string;
  status?: string;
  createdAt?: string;
  slaDeadline?: string;
  assignedTo?: string;
  items?: any[];
  context?: Record<string, any>;
}

export interface ApprovalResponse {
  requestId: string;
  respondedBy: string;
  respondedAt?: string;
  action: 'APPROVE' | 'APPROVE_WITH_EDITS' | 'REJECT' | 'ESCALATE' | 'DELEGATE' | 'REQUEST_INFO';
  comment?: string;
  delegateTo?: string;
}

export interface ControlAttestation {
  id?: string;
  controlId: string;
  workflowRunId?: string;
  evaluatedAt?: string;
  passed: boolean;
  expressionEvaluated?: string;
  inputValues?: Record<string, any>;
  inputHash?: string;
  evaluatedBy?: string;
}

export interface SpelResult {
  result: any;
  type: string;
  error: string | null;
}

export interface ExcelConversionResult {
  code: string;
  language: string;
  workbookName: string;
  inputCells: string[];
  outputCells: string[];
  formulaCount: number;
  dependencyCount: number;
}

export interface SubmissionManifest {
  id?: string;
  workflowRunId?: string;
  sourceRegion?: string;
  authoritativeFileId?: string;
  authoritativeFileName?: string;
  fileContentHash?: string;
  versionAssertedBy?: string;
}

// ─── Service ──────────────────────────────────────────────────────────────────

@Injectable({
  providedIn: 'root'
})
export class ProcessEngineService extends BaseService {

  private readonly baseUrl = `${this.backendUrl}/process`;

  constructor(private http: HttpClient) {
    super();
  }

  // ── Ontology ──────────────────────────────────────────────────────────────

  listOntologies(): Observable<OntologySchema[]> {
    return this.http.get<OntologySchema[]>(`${this.baseUrl}/ontology`);
  }

  createOntology(schema: OntologySchema): Observable<OntologySchema> {
    return this.http.post<OntologySchema>(`${this.baseUrl}/ontology`, schema);
  }

  getOntology(id: string, version: number = 1): Observable<OntologySchema> {
    const params = new HttpParams().set('version', version.toString());
    return this.http.get<OntologySchema>(`${this.baseUrl}/ontology/${id}`, { params });
  }

  updateOntology(id: string, schema: OntologySchema): Observable<OntologySchema> {
    return this.http.put<OntologySchema>(`${this.baseUrl}/ontology/${id}`, schema);
  }

  validateData(
    ontologyId: string,
    entityType: string,
    version: number,
    data: Record<string, any>
  ): Observable<ValidationRule[]> {
    const params = new HttpParams()
      .set('entityType', entityType)
      .set('version', version.toString());
    return this.http.post<ValidationRule[]>(
      `${this.baseUrl}/ontology/${ontologyId}/validate`,
      data,
      { params }
    );
  }

  // ── Process Definitions ───────────────────────────────────────────────────

  listProcessDefinitions(): Observable<ProcessDefinition[]> {
    return this.http.get<ProcessDefinition[]>(`${this.baseUrl}/definition`);
  }

  createProcess(definition: ProcessDefinition): Observable<ProcessDefinition> {
    return this.http.post<ProcessDefinition>(`${this.baseUrl}/definition`, definition);
  }

  getProcess(id: string, version: number = 1): Observable<ProcessDefinition> {
    const params = new HttpParams().set('version', version.toString());
    return this.http.get<ProcessDefinition>(`${this.baseUrl}/definition/${id}`, { params });
  }

  approveProcess(id: string, approvedBy: string): Observable<ProcessDefinition> {
    return this.http.post<ProcessDefinition>(
      `${this.baseUrl}/definition/${id}/approve`,
      { approvedBy }
    );
  }

  // ── Workflow Runs ─────────────────────────────────────────────────────────

  startRun(processDefinitionId: string, initialData?: Record<string, any>): Observable<WorkflowRun> {
    return this.http.post<WorkflowRun>(`${this.baseUrl}/run`, {
      processDefinitionId,
      initialData: initialData ?? {}
    });
  }

  getRun(id: string): Observable<WorkflowRun> {
    return this.http.get<WorkflowRun>(`${this.baseUrl}/run/${id}`);
  }

  listActiveRuns(): Observable<WorkflowRun[]> {
    return this.http.get<WorkflowRun[]>(`${this.baseUrl}/run/active`);
  }

  listAllRuns(): Observable<WorkflowRun[]> {
    return this.http.get<WorkflowRun[]>(`${this.baseUrl}/run/all`);
  }

  cancelRun(runId: string): Observable<WorkflowRun> {
    return this.http.post<WorkflowRun>(`${this.baseUrl}/run/${runId}/cancel`, {});
  }

  completeHumanStep(
    runId: string,
    stepId: string,
    completedBy: string,
    outputs?: Record<string, any>
  ): Observable<WorkflowRun> {
    return this.http.post<WorkflowRun>(
      `${this.baseUrl}/run/${runId}/complete-step`,
      { stepId, completedBy, outputs: outputs ?? {} }
    );
  }

  resumeAfterApproval(runId: string, response: ApprovalResponse): Observable<WorkflowRun> {
    return this.http.post<WorkflowRun>(`${this.baseUrl}/run/${runId}/resume`, response);
  }

  // ── Approvals ─────────────────────────────────────────────────────────────

  getPendingApprovals(assignedTo?: string): Observable<ApprovalRequest[]> {
    let params = new HttpParams();
    if (assignedTo) {
      params = params.set('assignedTo', assignedTo);
    }
    return this.http.get<ApprovalRequest[]>(`${this.baseUrl}/approval/pending`, { params });
  }

  getApprovalRequest(id: string): Observable<ApprovalRequest> {
    return this.http.get<ApprovalRequest>(`${this.baseUrl}/approval/${id}`);
  }

  submitApproval(id: string, response: ApprovalResponse): Observable<WorkflowRun> {
    return this.http.post<WorkflowRun>(`${this.baseUrl}/approval/${id}/respond`, response);
  }

  // ── Controls ──────────────────────────────────────────────────────────────

  createControl(control: ControlDefinition): Observable<ControlDefinition> {
    return this.http.post<ControlDefinition>(`${this.baseUrl}/control`, control);
  }

  listControls(): Observable<ControlDefinition[]> {
    return this.http.get<ControlDefinition[]>(`${this.baseUrl}/control`);
  }

  getControl(id: string): Observable<ControlDefinition> {
    return this.http.get<ControlDefinition>(`${this.baseUrl}/control/${id}`);
  }

  evaluateControl(
    controlId: string,
    runId: string,
    data: Record<string, any>
  ): Observable<ControlAttestation> {
    return this.http.post<ControlAttestation>(
      `${this.baseUrl}/control/${controlId}/evaluate`,
      { runId, data }
    );
  }

  getControlResults(runId: string): Observable<ControlAttestation[]> {
    return this.http.get<ControlAttestation[]>(`${this.baseUrl}/control/results/${runId}`);
  }

  // ── SpEL Evaluation ───────────────────────────────────────────────────────

  evaluateSpelExpression(expression: string, context: Record<string, any>): Observable<SpelResult> {
    return this.http.post<SpelResult>(`${this.baseUrl}/spel/evaluate`, { expression, context });
  }

  // ── Excel Formula Conversion / Execution ──────────────────────────────────

  convertExcelFormulas(spreadsheetGraphJson: string, targetLanguage = 'javascript'): Observable<ExcelConversionResult> {
    return this.http.post<ExcelConversionResult>(`${this.baseUrl}/excel/convert`, {
      spreadsheetGraphJson, targetLanguage
    });
  }

  executeExcelFormulas(
    spreadsheetGraphJson: string,
    cellOverrides?: Record<string, any>,
    targetLanguage = 'javascript',
    generatedCode?: string
  ): Observable<Record<string, any>> {
    return this.http.post<Record<string, any>>(`${this.baseUrl}/excel/execute`, {
      spreadsheetGraphJson, cellOverrides, targetLanguage, generatedCode
    });
  }

  // ── Manifests ─────────────────────────────────────────────────────────────

  createManifest(manifest: SubmissionManifest): Observable<SubmissionManifest> {
    return this.http.post<SubmissionManifest>(`${this.baseUrl}/manifest`, manifest);
  }

  assertAuthoritativeVersion(
    manifestId: string,
    fileId: string,
    assertedBy: string
  ): Observable<SubmissionManifest> {
    return this.http.post<SubmissionManifest>(
      `${this.baseUrl}/manifest/${manifestId}/assert-version`,
      { fileId, assertedBy }
    );
  }

  // ── Process Discovery ─────────────────────────────────────────────────────

  discoverProcesses(graphNodeIds?: string[], options?: any): Observable<any> {
    return this.http.post<any>(`${this.backendUrl}/process/discovery/suggest`, { graphNodeIds, options });
  }

  analyzeEmailFlows(graphNodeIds?: string[]): Observable<any> {
    return this.http.post<any>(`${this.backendUrl}/process/discovery/email-flows`, { graphNodeIds });
  }

  analyzeExcelFlows(graphNodeIds?: string[]): Observable<any> {
    return this.http.post<any>(`${this.backendUrl}/process/discovery/excel-flows`, { graphNodeIds });
  }

  acceptSuggestion(suggestion: any): Observable<any> {
    return this.http.post<any>(`${this.backendUrl}/process/discovery/accept`, suggestion);
  }

  analyzeDocumentFlows(graphNodeIds?: string[]): Observable<any> {
    return this.http.post<any>(`${this.backendUrl}/process/discovery/document-flows`, { graphNodeIds });
  }

  analyzeCrossDocumentFlows(graphNodeIds?: string[]): Observable<any> {
    return this.http.post<any>(`${this.backendUrl}/process/discovery/cross-document-flows`, { graphNodeIds });
  }

  // ── Integration Discovery ─────────────────────────────────────────────────

  listIntegrations(): Observable<any> {
    return this.http.get<any>(`${this.baseUrl}/integrations`);
  }

  listStepTypes(): Observable<Array<{ type: string; description: string }>> {
    return this.http.get<Array<{ type: string; description: string }>>(`${this.baseUrl}/step-types`);
  }

  // ── Run Inspection ────────────────────────────────────────────────────────

  getRunSteps(runId: string): Observable<any> {
    return this.http.get<any>(`${this.baseUrl}/run/${runId}/steps`);
  }

  getRunContext(runId: string): Observable<any> {
    return this.http.get<any>(`${this.baseUrl}/run/${runId}/context`);
  }

  getStepResult(runId: string, stepId: string): Observable<any> {
    return this.http.get<any>(`${this.baseUrl}/run/${runId}/step/${stepId}`);
  }

  updateRunData(runId: string, data: Record<string, any>): Observable<any> {
    return this.http.put<any>(`${this.baseUrl}/run/${runId}/data`, data);
  }

  // ── Email Value Extraction ────────────────────────────────────────────────

  extractEmailValues(emailBody: string, subject?: string, messageId?: string): Observable<any> {
    return this.http.post<any>(`${this.backendUrl}/email/extract-values/extract`, {
      body: emailBody, subject, messageId
    });
  }

  mapEmailValuesToCells(emailBody: string, spreadsheetGraphJson: string): Observable<any> {
    return this.http.post<any>(`${this.backendUrl}/email/extract-values/map-to-cells`, {
      body: emailBody, spreadsheetGraphJson
    });
  }
}
